Added `miri` and it detected UB:

```
Run cd runtime
MIRIFLAGS="-Zmiri-disable-isolation" cargo miri nextest run --lib utils::buffer::paged::append::tests:: --partition hash:2/8
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.26s
info: for the target platform, using target runner `/home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/bin/cargo-miri runner` defined by `target.cfg(all()).runner` specified by `--config`
info: for the host platform, using target runner `/home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/bin/cargo-miri runner` defined by `target.cfg(all()).runner` specified by `--config`
────────────
 Nextest run ID 0bd21d87-e848-4695-a5c6-550c2836b9e6 with nextest profile: default-miri
    Starting 5 tests across 27 binaries (4851 tests skipped)
        FAIL [  14.416s] (1/5) commonware-runtime utils::buffer::paged::append::tests::test_read_many_into_matches_read_at
  stdout ───

    running 1 test
    test utils::buffer::paged::append::tests::test_read_many_into_matches_read_at ... 
  stderr ───
    error: Undefined Behavior: trying to retag from <911647> for SharedReadOnly permission at alloc310086[0x0], but that tag does not exist in the borrow stack for this location
        --> runtime/src/utils/buffer/paged/cache.rs:249:33
         |
     249 |             let mut remaining = buf.len();
         |                                 ^^^ this error occurs as part of retag at alloc310086[0x0..0xa]
         |
         = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
         = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
    help: <911647> was created by a Unique retag at offsets [0x0..0xa]
        --> runtime/src/utils/buffer/paged/append.rs:498:17
         |
     498 |                 cache_ranges.push((slice, offset));
         |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    help: <911647> was later invalidated at offsets [0x0..0x15e] by a Unique function-entry retag inside this call
        --> runtime/src/utils/buffer/paged/append.rs:494:36
         |
     494 |                 let ptr = unsafe { buf.as_mut_ptr().add(i * item_size) };
         |                                    ^^^^^^^^^^^^^^^^
         = note: BACKTRACE (of the first span) on thread `utils::buffer::`:
         = note: inside closure at runtime/src/utils/buffer/paged/cache.rs:249:33: 249:36
         = note: inside `std::vec::Vec::<T, A>::retain_mut::process_loop::<{closure@runtime/src/utils/buffer/paged/cache.rs:248:27: 248:50}, (&mut [u8], u64), std::alloc::Global, false>` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:2335:21: 2335:27
         = note: inside `std::vec::Vec::<(&mut [u8], u64)>::retain_mut::<{closure@runtime/src/utils/buffer/paged/cache.rs:248:27: 248:50}>` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:2361:9: 2361:69
    note: inside `utils::buffer::paged::cache::CacheRef::read_cached_many`
        --> runtime/src/utils/buffer/paged/cache.rs:248:9
         |
     248 | /         ranges.retain_mut(|(buf, logical_offset)| {
     249 | |             let mut remaining = buf.len();
     250 | |             let mut offset = *logical_offset;
     251 | |             let mut dst = 0;
    ...    |
     263 | |             remaining > 0
     264 | |         });
         | |__________^
    note: inside closure
        --> runtime/src/utils/buffer/paged/append.rs:526:9
         |
     526 |         self.cache_ref.read_cached_many(self.id, &mut cache_ranges);
         |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    note: inside closure
        --> runtime/src/utils/buffer/paged/append.rs:1148:18
         |
    1148 |                 .await
         |                  ^^^^^
         = note: inside `<std::pin::Pin<&mut {async block@runtime/src/utils/buffer/paged/append.rs:1123:58: 1123:68}> as futures::Future>::poll` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/future.rs:133:9: 133:61
         = note: inside `<std::pin::Pin<&mut {async block@runtime/src/utils/buffer/paged/append.rs:1123:58: 1123:68}> as futures::FutureExt>::poll_unpin` at /mnt/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/futures-util-0.3.31/src/future/future/mod.rs:558:9: 558:32
         = note: inside `<futures::future::Select<std::pin::Pin<&mut tokio::sync::oneshot::Receiver<std::boxed::Box<dyn std::any::Any + std::marker::Send>>>, std::pin::Pin<&mut {async block@runtime/src/utils/buffer/paged/append.rs:1123:58: 1123:68}>> as futures::Future>::poll` at /mnt/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/futures-util-0.3.31/src/future/select.rs:118:35: 118:51
    note: inside closure
        --> runtime/src/utils/handle.rs:237:38
         |
     237 |         match select(panicked, task).await {
         |                                      ^^^^^
    note: inside closure
        --> runtime/src/deterministic.rs:618:54
         |
     618 |                         if let Poll::Ready(result) = root.as_mut().poll(&mut cx) {
         |                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
         = note: inside `<{closure@runtime/src/deterministic.rs:560:52: 560:54} as std::ops::FnOnce<()>>::call_once - shim` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71
         = note: inside `<std::panic::AssertUnwindSafe<{closure@runtime/src/deterministic.rs:560:52: 560:54}> as std::ops::FnOnce<()>>::call_once` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:274:9: 274:19
         = note: inside `std::panicking::catch_unwind::do_call::<std::panic::AssertUnwindSafe<{closure@runtime/src/deterministic.rs:560:52: 560:54}>, ()>` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:582:40: 582:43
         = note: inside `std::panicking::catch_unwind::<(), std::panic::AssertUnwindSafe<{closure@runtime/src/deterministic.rs:560:52: 560:54}>>` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:545:19: 545:88
         = note: inside `std::panic::catch_unwind::<std::panic::AssertUnwindSafe<{closure@runtime/src/deterministic.rs:560:52: 560:54}>, ()>` at /home/runner/.rustup/toolchains/nightly-2025-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:359:14: 359:40
    note: inside `deterministic::Runner::start_and_recover::<{closure@runtime/src/utils/buffer/paged/append.rs:1123:24: 1123:57}, {async block@runtime/src/utils/buffer/paged/append.rs:1123:58: 1123:68}>`
        --> runtime/src/deterministic.rs:560:22
         |
     560 |           let result = catch_unwind(AssertUnwindSafe(|| loop {
         |  ______________________^
     561 | |             // Ensure we have not exceeded our deadline
     562 | |             {
     563 | |                 let current = executor.time.lock();
    ...    |
     665 | |             executor.metrics.iterations.inc();
     666 | |         }));
         | |___________^
    note: inside `<deterministic::Runner as Runner>::start::<{closure@runtime/src/utils/buffer/paged/append.rs:1123:24: 1123:57}, {async block@runtime/src/utils/buffer/paged/append.rs:1123:58: 1123:68}>`
        --> runtime/src/deterministic.rs:736:27
         |
     736 |         let (output, _) = self.start_and_recover(f);
         |                           ^^^^^^^^^^^^^^^^^^^^^^^^^
    note: inside closure
        --> runtime/src/utils/buffer/paged/append.rs:1123:9
         |
    1123 | /         executor.start(|context: deterministic::Context| async move {
    1124 | |             let (blob, blob_size) = context.open("test_partition", b"rmany").await.unwrap();
    1125 | |             let cache_ref = CacheRef::from_pooler(
    1126 | |                 &context.with_label("cache"),
    ...    |
    1160 | |         });
         | |__________^
         = note: inside `tracing::dispatcher::with_default::<(), {closure@runtime/src/utils/buffer/paged/append.rs:1119:5: 1119:28}>` at /mnt/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tracing-core-0.1.34/src/dispatcher.rs:262:5: 262:8
    note: inside `utils::buffer::paged::append::tests::test_read_many_into_matches_read_at`
        --> runtime/src/utils/buffer/paged/append.rs:1119:5
         |
    1119 |     #[test_traced("DEBUG")]
         |     ^^^^^^^^^^^^^^^^^^^^^^^
    note: inside closure
        --> runtime/src/utils/buffer/paged/append.rs:1120:45
         |
    1119 |     #[test_traced("DEBUG")]
         |     ----------------------- in this attribute macro expansion
    1120 |     fn test_read_many_into_matches_read_at() {
         |                                             ^
         = note: this error originates in the attribute macro `test_traced` (in Nightly builds, run with -Z macro-backtrace for more info)

    note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

    error: aborting due to 1 previous error
    

    (test failed with exit code 1)

  Cancelling due to test failure: 3 tests still running
        PASS [  15.957s] (2/5) commonware-runtime utils::buffer::paged::append::tests::test_corrupted_crc_both_slots_len_too_large
        PASS [  16.286s] (3/5) commonware-runtime utils::buffer::paged::append::tests::test_data_prefix_not_overwritten
        PASS [  19.189s] (4/5) commonware-runtime utils::buffer::paged::append::tests::test_crc_slot_protection_across_page_boundary
────────────
     Summary [  19.192s] 4/5 tests run: 3 passed, 1 failed, 4851 skipped
        FAIL [  14.416s] (1/5) commonware-runtime utils::buffer::paged::append::tests::test_read_many_into_matches_read_at
warning: 1/5 tests were not run due to test failure (run with --no-fail-fast to run all tests, or run with --max-fail)
error: test run failed
error: Recipe `miri` failed on line 93 with exit code 100
Error: Process completed with exit code 100.
```

Pushing a fix now.

--- a/.github/workflows/slow.yml
+++ b/.github/workflows/slow.yml
@@ -200,7 +200,7 @@ jobs:
       run: |
         cd runtime
         just miri iobuf:: --partition hash:${{matrix.partition}}/8
-    - name: Run miri tests for runtime::paged::append
+    - name: Run miri tests for runtime::buffer::paged::append
       run: |
         cd runtime
         just miri utils::buffer::paged::append::tests:: --partition hash:${{matrix.partition}}/8

--- a/runtime/src/utils/buffer/paged/append.rs
+++ b/runtime/src/utils/buffer/paged/append.rs
@@ -463,7 +463,14 @@ impl<B: Blob> Append<B> {
         offsets: &[u64],
         item_size: usize,
     ) -> Result<(), Error> {
-        debug_assert_eq!(buf.len(), offsets.len() * item_size);
+        assert_eq!(
+            buf.len(),
+            offsets
+                .len()
+                .checked_mul(item_size)
+                .expect("read_many_into buffer length overflow"),
+            "read_many_into requires buf.len() == offsets.len() * item_size"
+        );
         if offsets.is_empty() {
             return Ok(());
         }
@@ -1103,6 +1110,30 @@ mod tests {
             let mut buf = vec![0u8; 8];
             append.read_many_into(&mut buf, &[0], 8).await.unwrap();
             assert_eq!(&buf, &data);
+        });
+    }
+
+    #[test]
+    #[should_panic(expected = "read_many_into requires buf.len() == offsets.len() * item_size")]
+    fn test_read_many_into_short_buffer_panics() {
+        let executor = deterministic::Runner::default();
+        executor.start(|context: deterministic::Context| async move {
+            let (blob, blob_size) = context.open("test_partition", b"rmany").await.unwrap();
+            let cache_ref = CacheRef::from_pooler(
+                &context.with_label("cache"),
+                PAGE_SIZE,
+                NZUsize!(BUFFER_SIZE),
+            );
+            let append = Append::new(blob, blob_size, BUFFER_SIZE, cache_ref)
+                .await
+                .unwrap();
+
+            let data: Vec<u8> = (0..16).collect();
+            append.append(&data).await.unwrap();
+
+            let offsets = [0u64, 4];
+            let mut buf = vec![0u8; 7];
+            append.read_many_into(&mut buf, &offsets, 4).await.unwrap();
         });
     }
 

--- a/runtime/src/utils/buffer/paged/append.rs
+++ b/runtime/src/utils/buffer/paged/append.rs
@@ -478,40 +478,30 @@ impl<B: Blob> Append<B> {
             return Err(Error::BlobInsufficientLength);
         }
 
-        // Iterate over requested offsets and copy items that overlap with the tip
-        // buffer directly into the output. Items fully or partially below the tip
-        // need cache/blob reads and are recorded as (slice, offset) pairs. Slices
-        // are built via raw pointer arithmetic since each item occupies a disjoint,
-        // item_size-aligned region of the output.
+        // Iterate over fixed-size output slots and copy items that overlap with the
+        // tip buffer directly into place. Items fully or partially below the tip
+        // need cache/blob reads and are recorded as (slice, offset) pairs.
+        // `chunks_exact_mut` yields disjoint per-item slots, so we never have to
+        // reborrow the parent buffer while cache/blob destinations remain live.
+        if item_size == 0 {
+            return Ok(());
+        }
         let mut cache_ranges: Vec<(&mut [u8], u64)> = Vec::new();
-        for (i, &offset) in offsets.iter().enumerate() {
+        for (item_buf, &offset) in buf.chunks_exact_mut(item_size).zip(offsets.iter()) {
             let end = offset + item_size as u64;
 
             if end <= buffer.offset {
                 // Entirely below tip -- needs cache read.
-                // SAFETY: i * item_size < buf.len() because i < offsets.len() and
-                // buf.len() == offsets.len() * item_size.
-                let ptr = unsafe { buf.as_mut_ptr().add(i * item_size) };
-                // SAFETY: Each i is unique so the resulting slices are
-                // non-overlapping, and item_size keeps each within its slot.
-                let slice = unsafe { core::slice::from_raw_parts_mut(ptr, item_size) };
-                cache_ranges.push((slice, offset));
+                cache_ranges.push((item_buf, offset));
             } else if offset >= buffer.offset {
                 // Entirely in tip buffer.
                 let src = (offset - buffer.offset) as usize;
-                let item_buf = &mut buf[i * item_size..(i + 1) * item_size];
                 item_buf.copy_from_slice(&buffer.as_ref()[src..src + item_size]);
             } else {
                 // Straddles tip boundary: copy suffix from tip, record prefix for cache.
                 let prefix_len = (buffer.offset - offset) as usize;
-                let item_suffix_buf = &mut buf[i * item_size + prefix_len..(i + 1) * item_size];
-                item_suffix_buf.copy_from_slice(&buffer.as_ref()[..item_size - prefix_len]);
-                // SAFETY: i * item_size < buf.len() (same as above).
-                let ptr = unsafe { buf.as_mut_ptr().add(i * item_size) };
-                // SAFETY: Each i is unique so slices are non-overlapping,
-                // and prefix_len <= item_size keeps within the slot.
-                let slice = unsafe { core::slice::from_raw_parts_mut(ptr, prefix_len) };
-                cache_ranges.push((slice, offset));
+                item_buf[prefix_len..].copy_from_slice(&buffer.as_ref()[..item_size - prefix_len]);
+                cache_ranges.push((&mut item_buf[..prefix_len], offset));
             }
         }
 

